### DISCUSSION TITLE
How to plot multiple functions efficiently over the same points

### DISCUSSION THREAD

**Discussion by faze-geek** [sympy/sympy#25183] [Q&A]
I am trying to relate two functions. For that I was using `LineOver1DRangeSeries` from the plotting module, since it supports an effective sampling. 

```
     x, y = LineOver1DRangeSeries(func1,(_w, lower, upper), xscale='log', **kwargs).get_points()
    print(x)
    print(y)

    x, y = LineOver1DRangeSeries(func2,(_w, lower, upper), xscale='log', **kwargs).get_points()
    print(x)
    print(y)

[1e-05, 1.418281451544817e-05, 2.11386638818207e-05, 2.9558256262855514e-05, 4.452646718643275e-05....
[9.542425093862445, 9.542425093325525, 9.542425092021384, 9.54242508975566, 9.542425083869489,......

[1e-05, 1.5158439697186553e-05, 2.1636724316035136e-05, 3.0147800700320413e-05, 4.3919640587422284e-05....
[0.00019098593168410233, 0.00028950487279295867, 0.0004132309950005295, 0.0005757805798594919,....
```
As of now I can't relate `func1 `and `func2` because if you see the printed `x `values they are not the same even though the scale is `log `and my `lower `and `upper `bounds are the same.

Can someone help to relate my functions using same set of points from LineOver1DRangeSeries ?

**Comment by faze-geek** [comment]
@Davide-sd @sylee957 
Could you help me out here. Thanks in advance !

**Comment by Davide-sd** [comment]
@faze-geek are you using sympy plotting module? If so, try to set "adaptive=False".

**Reply by faze-geek** [reply]
Yup. Let me try and get back to you.

**Reply by faze-geek** [reply]
```
    x, y = LineOver1DRangeSeries(func1,(_w, lower, upper), xscale='log', adaptive= 'False', **kwargs).get_points()
    print(x)
   
    x, y = LineOver1DRangeSeries(func2, (_w,lower, upper), xscale='log',adaptive= 'False', **kwargs).get_points()
    print(x)
   

[1e-05, 1.4206182448514702e-05, 2.0852885442321886e-05, 2.8407235199884208e-05, 3.877128536600936e-05, 6.155737039892703e-05,.....

[1e-05, 1.3833993100872492e-05, 2.0491203991955066e-05, 2.79679664954481e-05, 3.715699534583969e-05, 5.350319012888729e-05,.....
```

**Reply by faze-geek** [reply]
The x values are still not the same it seems !

**Reply by Davide-sd** [reply]
You have to use `adaptive=False`. It must be a boolean, not a string!!!

**Reply by faze-geek** [reply]
I would want them to be same, so that the points I get for relating the y's of both are correct. I would like to use a symbolic approach instead of `np.logspace`

**Reply by faze-geek** [reply]
> You have to use `adaptive=False`. It must be a boolean, not a string!!!

Oops :)
```
It leads to a new problem tho -
x,y = LineOver1DRangeSeries(fucn1, (_w,lower,upper), xscale='log',adaptive= False,nb_of_points =100, **kwargs).get_points()

[1.00002303        inf        inf        inf        inf        inf
        inf        inf        inf        inf        inf        inf
        inf        inf        inf        inf        inf        inf
        inf        inf        inf        inf        inf        inf
        inf        inf        inf        inf        inf        inf
        inf        inf        inf        inf        inf        inf
        inf        inf        inf        inf        inf        inf
        inf        inf        inf        inf        inf        inf
        inf        inf        inf        inf        inf        inf
        inf        inf        inf        inf        inf        inf
        inf        inf        inf        inf        inf        inf
        inf        inf        inf        inf        inf        inf
        inf        inf        inf        inf        inf        inf
        inf        inf        inf        inf        inf        inf
        inf        inf        inf        inf        inf        inf
        inf        inf        inf        inf        inf        inf
        inf        inf        inf        inf]
```

**Reply by faze-geek** [reply]
Is this expected @Davide-sd ?

**Reply by Davide-sd** [reply]
I can't help you if you don't show `fucn1`...

**Reply by faze-geek** [reply]
```
     _w = Dummy("w", real=True)
    repl = I*_w
    w_expr = expr.subs({system.var: repl})

    func1 = -20*log(Abs(w_expr), 10)
    func2 = arg(w_expr)*180/pi
```

**Reply by faze-geek** [reply]
These are basically `magnitude `and `phase `functions in signals and systems theory which I am trying to relate for some context

**Reply by Davide-sd** [reply]
Can't construct the expression. There are missing variables... Please, do something like this:

```py
print_python(func1)
print_python(func2)
```

Copy the output here.

**Reply by faze-geek** [reply]
```
def mag_to_phase(system, initial_exp=-5, final_exp=5, **kwargs):
    expr = system.to_expr()

    _w = Dummy("w", real=True)
    repl = I*_w
    w_expr = expr.subs({system.var: repl})

    mag = -20*log(Abs(w_expr), 10)
    phase = arg(w_expr)*180/pi
    
    print_python(mag)
    print_python(phase)

mag_to_phase(TransferFunction(s+1,s**2+2*s+3,s))
e = -20*log(sqrt(_w**2 + 1)/sqrt(_w**4 - 2*_w**2 + 9))/log(10)
e = 180*arg((_w*I + 1)/(-_w**2 + 2*_w*I + 3))/pi
```

**Reply by faze-geek** [reply]
The functions I have to relate are above. Thanks for your patience.

**Reply by Davide-sd** [reply]
There must be some bug with the uniform evaluation strategy of the current plotting module in sympy.

Since you are interested in plotting, [this module that I created](https://sympy-plot-backends.readthedocs.io/en/latest/install.html) is able to plot it:

```py
from spb import plot, plotgrid
w = list(e1.free_symbols)[0]
n = 1000
p1 = plot(e1, (w, 1e-05, 1e05), xscale="log", adaptive=False, n=n, show=False, ylabel="Mag")
p2 = plot(e2, (w, 1e-05, 1e05), xscale="log", adaptive=False, n=n, show=False, ylabel="Arg")
plotgrid(p1, p2)
```

![image](https://github.com/sympy/sympy/assets/9921777/e32745f7-6ca6-492f-8eb1-9bc14aedf5db)

**Reply by Davide-sd** [reply]
@faze-geek did you give it a try? I know it's not part of sympy yet, but hopefully one day it will...

**Reply by faze-geek** [reply]
Hey . I will try it today for sure. For now I successfully did the sampling using `np.logspace` instead of `LineOver1DRangeSeries()` with `xscale = 'log'. ` . 
A symbolic method is always preferred over a numeric one unless accuracy is not sacrificed.

**Comment by sylee957** [comment]
Why do you need to align same x points for different functions? I don't think that the plotting module supports such parallel sampling yet, but the possible workaround is to use interpolation libraries like `scipy.interpolate` to fill out the data for mismatching points

**Comment by sylee957** [comment]
I think that the lower and upper ranges should be `-5, 5` instead of `1e-5, 1e5` if you use logscale sampling
It is weird semantics, but I'll try to look at the PR to see why does that implement that way

**Reply by faze-geek** [reply]
Yes I have taken care of that. Thanks !

**Reply by sylee957** [reply]
I already identify the issue with range a bug than feature becacuse there is inconsistency
https://github.com/sympy/sympy/issues/25190

**Comment by faze-geek** [comment]
Hi @Davide-sd , the code snippet you gave works well.
I have one more doubt, are you aware of any `sympy` or `matplotlib` routines which add an arrow on the trajectory of the curve. Something like this (in MATLAB taken from stackoverflow) .

![image](https://github.com/sympy/sympy/assets/90216905/647621bb-28a6-4726-8edc-1cb69aeb9c25)

P.S.- My apologies for going out of topic of discussion. This is also something which is relevant to the work I am doing and opening another discussion everytime does not seem feasible.

**Reply by Davide-sd** [reply]
I think you have to play with Matplotlib's [`FancyArrowPatch`](https://matplotlib.org/stable/api/_as_gen/matplotlib.patches.FancyArrowPatch.html). In this example, I'm adding an arrow at the end of a curve, which is built with an object of type `Path`.

```py
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches
import matplotlib.path as mpath

xx = np.linspace(0, 2*np.pi, 200)
yy = np.cos(xx)

fig, ax = plt.subplots()
ax.plot(xx, yy, ":", zorder=0)
path = mpath.Path(np.stack([xx[:100], yy[:100]]).T)
arrow = mpatches.FancyArrowPatch(path=path, arrowstyle="-|>",
        mutation_scale=15, color="r", zorder=1, linewidth=2)
ax.add_patch(arrow)
plt.show()
```

![image](https://github.com/sympy/sympy/assets/9921777/4de7f341-5b1a-4e4d-b791-4be3a301790c)


It is also possible to add multiple arrows, just provide `posA` and `posB` arguments instead of `Path`...
